pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+            backup_storage_provider_id: bsp_id,
+            rate,
+        });
+
+        // Return a successful DispatchResult
+        Ok(())
+    }
+
+    fn update_payment_stream(
+        bsp_account: &Self::AccountId,
+        user_account: &Self::AccountId,
+        new_rate: <Self::Balance as Inspect<Self::AccountId>>::Balance,
+    ) -> DispatchResult {
+        // Execute the logic to update a payment stream
+        Self::do_update_payment_stream(bsp_account, user_account, new_rate)?;
+
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Emit the corresponding event
+        Self::deposit_event(Event::<T>::PaymentStreamUpdated {
+            user_account: user_account.clone(),
+            backup_storage_provider_id: bsp_id,
+            new_rate,
+        });
+
+        // Return a successful DispatchResult
+        Ok(())
+    }
+
+    fn delete_payment_stream(
+        bsp_account: &Self::AccountId,
+        user_account: &Self::AccountId,
+    ) -> DispatchResult {
+        // Execute the logic to delete a payment stream
+        Self::do_delete_payment_stream(bsp_account, user_account)?;
+
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Emit the corresponding event
+        Self::deposit_event(Event::<T>::PaymentStreamDeleted {
+            user_account: user_account.clone(),
+            backup_storage_provider_id: bsp_id,
+        });
+
+        // Return a successful DispatchResult
+        Ok(())
+    }
+
+    fn update_last_valid_proof(
+        bsp_account: &Self::AccountId,
+        user_account: &Self::AccountId,
+        last_valid_proof_block: Self::BlockNumber,
+    ) -> DispatchResult {
+        // Ensure that the last valid proof block that is being submitted is not greater than the current block number
+        ensure!(
+            last_valid_proof_block <= frame_system::Pallet::<T>::block_number(),
+            Error::<T>::InvalidLastValidProofBlockNumber
+        );
+
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Get the information of the payment stream to update
+        let payment_stream = PaymentStreams::<T>::get(bsp_id, user_account)
+            .ok_or(Error::<T>::PaymentStreamNotFound)?;
+
+        // Ensure that the new last valid proof block is greater than the last valid proof block of the payment stream
+        ensure!(
+            last_valid_proof_block > payment_stream.last_valid_proof,
```

Comment:
Should we also check that it's greater than `last_charged_proof`? We could check it with `expect_or_err!` as `last_checked_proof` should always be less or equal than `last_valid_proof`.